service/automatic_run_experiment.py

```diff
@@ -62,26 +73,128 @@
-    fuzzers = experiment_changes.get_fuzzers_changed_since_last()
-    logs.info('Running experiment with fuzzers: %s.', ' '.join(fuzzers))
+def _get_experiment_name(experiment_config: dict) -> str:
+    """Returns the name of the experiment described by experiment_config as a
+    string."""
+    # Use str because the yaml parser will parse things like `2020-05-06` as
+    # a datetime if not included in quotes.
+    return str(experiment_config['experiment'])
+
+
+def validate_experiment_name(experiment_name):
+    """Returns True if |experiment_name| is valid."""
+    return EXPERIMENT_NAME_REGEX.match(experiment_name) is not None
+
+
+def validate_experiment_requests(experiment_requests):
+    """Returns True if all requests in |experiment_requests| are valid."""
+    # This function tries to find as many requests as possible.
+    if PAUSE_SERVICE_KEYWORD in experiment_requests:
+        # This is a special case where a string is used instead of an experiment
+        # to tell the service not to run experiments automatically. Remove it
+        # from the list because it fails validation.
+        experiment_requests = experiment_requests[:]  # Don't mutate input.
+        experiment_requests.remove(PAUSE_SERVICE_KEYWORD)
+
+    all_fuzzers = set(fuzzer_utils.get_fuzzer_names())
+    valid = True
+    # Validate format.
+    for request in experiment_requests:
+        if not isinstance(request, dict):
+            logger.error('Request: %s is not a dict.', request)
+            experiment_requests.remove(request)
+            valid = False
+            continue
+
+        if 'experiment' not in request:
+            logger.error('Request: %s does not have field "experiment".',
+                         request)
+            valid = False
+            continue
+
+        experiment = _get_experiment_name(request)
+        if not validate_experiment_name(experiment):
+            valid = False
+            logger.error('Experiment name: %s is not valid.', experiment)
+            # Request isn't so malformed that we can finding other issues.
+            # if present. Don't continue.
+
+        experiment = request['experiment']
+        if not request.get('fuzzers'):
+            logger.error('Request: %s does not specify any fuzzers.', request)
+            valid = False
+            continue
+
+        experiment_fuzzers = request['fuzzers']
+        for fuzzer in experiment_fuzzers:
+            if fuzzer in all_fuzzers:
+                continue
+            # Fuzzer isn't valid.
+            logger.error('Fuzzer: %s in experiment %s is not valid.', fuzzer,
+                         experiment)
+            valid = False
+
+    if not valid:
+        # Don't try the next validation step if the previous failed, we might
+        # exception.
+        return valid
+
+    # Make sure experiment requests have a unique name, we can't run the same
+    # experiment twice.
+    counts = collections.Counter(
```

Comment:
Why not just keep a list of experiment name inside the loop above and do validation there instead of another loop.

Reply:
I think that has n^2 complexity right? The list will probably be small enough that it doesn't matter, but that is the reason I didn't do this.